Add support of interface created by dispatch script

cathay4t opened the pull request:
Example YAML:

```yml
dispatch:
  interface-types:
  - type: vxcan
    activation: ip link add $name type vxcan peer $peer
    deactivation: ip link del $name
    kernel-index-getter: cat /sys/class/net/$name/ifindex
    allowed-variable-names:
    - peer
interfaces:
- name: vxcan1
  description: vxcan interface created by dispatch script
  type: dispatch
  state: up
  ipv4:
    enabled: true
    address:
    - ip: 192.0.2.1
      prefix-length: 24
  dispatch:
    type: vxcan
    variables:
      peer: vxcan1-ep
```

To remove a interface type, please set `state: absent` to interface type
in `dispatch.interface-types`.

Example YAML included as `examples/dispatch_iface.yml`.

Integration and unit test cases included.

cathay4t commented:
To try it out, you may use NetworkManager from https://people.redhat.com/fge/user_defined_iface/

cathay4t commented:
This is not about adding support of vxcan using `user-defined` interface.

The vxcan here is example showing we can use this `user-defined` section to support arbitrary interface set by user-defined script even that is not supported by NetworkManager or nmstate natively.

qinqon commented:
the "type" should denote that this is a custom thing 
```yaml
custom-devices:
  - name: vxcan
    handler-script: |
      #!/bin/bash
      ifname=$1
      action=$2

      if [ "$action" = "device-add" ]; then
          peer=$CONNECTION_USER_VXCAN__PEER

          if [ -z "$peer" ]; then
              echo "ERROR=Missing peer name"
              exit 1
          fi

          if ! err=$(ip link add "$ifname" type vxcan peer "$peer" 2>&1); then
              echo "ERROR=Failed creating the interface: $err"
              exit 2
          fi
          ip link set $peer up

          echo IFINDEX="$(cat /sys/class/net/"$ifname"/ifindex)"
          exit 0
      elif [ "$action" = "device-delete" ]; then
          # Delete the interface created by "device-add" here.
          ip link del "$ifname"
          exit 0
      fi

interfaces:
  - name: vxcan1
    type: custom-device
    description: "user defined vxcan interface"
    custom-device:
      name: vxcan
      config:
        peer: "vxcan1-ep"
    ipv4:
      enabled: true
      address:
        - ip: 192.0.2.1
          prefix-length: 24
```

cathay4t commented:
> the "type" should denote that this is a custom thing

Nmstate has no intention allowing user to override native interface types, and we have check make sure user defined interface type is valid. So nmstate is smart enough to know user is referring a custom interface.

cathay4t commented:
NetworkManager patch been merged. Please review nmstate part.

tyll commented on examples/user_defined_iface.yml:
```diff
@@ -0,0 +1,43 @@
+
+interfaces:
+    - name: vxcan1
+      type: vxcan
```
IMHO user-defined types need their own namespace, otherwise we will have a hard time in the future to add new internal interface types. Maybe something like: `type: generic: vxcan`.

cathay4t replied:
`generic` is not explicit enough, many thing can fit into it. I think `user-defined` is clear enough without any document explain what it is.

tyll replied:
My concern here is that `type: vxcan` needs to have it's own namespace because this way user defined types will collide with possible type we will define in the the future in core Nmstate. So with `user-defined`, it would be something like `type: user-defined: vxcan`. I am more in favor of `type: plugin: vxcan` now. Another option would be to use `type: plugin` and move all plugin details to `plugin-options`. The type can then be optional and the presence of `plugin-options` could be enough to identify this as type `plugin`.:

```yaml
- name: vxcan1
  type: plugin
  plugin:
      type: vxcan
      state:
         peer: "vxcan1-ep"
```

EDIT: Removed the `-options` suffix, since it is not neccessary.

tyll commented on examples/user_defined_iface.yml:
```diff
@@ -0,0 +1,43 @@
+---
+user-defined:
```
In NM we call these "generic", what's the reason to use a different term in Nmstate?

tyll replied:
If the items have `type: interface`, then `generic-types` could be an alternative name for `user-defined`:

```
generic-types:
    - type: interface
      name: vxcan
      handler-scripts: ...
```

cathay4t replied:
The `generic-types` cannot explains itself without document, but `user-defined` can.

tyll replied:
Making it self-explanatory is a good idea, I don't see this with `user-defined` as well, though. In general the whole desired state is "user-defined", since it is what the user defines as the desired state.

How about using something like `plugins:` or `addons:` instead?

tyll commented on examples/user_defined_iface.yml:
```diff
@@ -0,0 +1,43 @@
+---
+user-defined:
+    interface-types:
```
What's the reason for the extra level? Adding `type: interface` as a property to the items would be an alternative to declare that this is a custom interface type.

cathay4t replied:
Reserve for future `user-defined` stuff.

tyll replied:
You can still define other plugins with this as well by using different types:

```
plugins:
    - type: interface
      name: vxcan
      handler: ....
   - type: routes
     name: blackhole
     handler: ...
```

tyll replied:
If you prefer the extra level, still, something like

```
plugins:
    interfaces:
        - name: vxcan
           handler: ...
    routes:
        - name: blackhole
          handler: ...
```

is also something I can accept, I guess.

tyll commented on examples/user_defined_iface.yml:
```diff
@@ -0,0 +1,43 @@
+            ifname=$1
+            action=$2
+
+            if [ "$action" = "device-add" ]; then
```
The action is called "device-add" but we call devices "interfaces" elsewhere. Also, what is the reason to call it action instead of `state` and then using the same values for state as we allow in the list of interfaces?

tyll replied:
Ah, it is a dispatcher script.

tyll commented on rust/src/lib/ifaces/user_defined.rs:
```diff
@@ -0,0 +1,139 @@
+                                ErrorKind::InvalidArgument,
+                                format!(
+                                    "Interface type {kind} is \
+                                not native supported or defined in \
+                                user-defined section"
```
```suggestion
                                not natively supported or defined in \
                                a user-defined section"
```

The own namespace for generic types would also help to clarify this error message.

tyll commented on examples/user_defined_iface.yml:
```diff
@@ -0,0 +1,43 @@
+
+                if [ -z "$peer" ]; then
+                    echo "ERROR=Missing peer name"
+                    exit 1
```
Is the meaning of the error codes defined somewhere? This is more like an observation for NM: It seems it should define different return codes such as $NM_RETURN_MISSING_PARAMETER or $NM_RETURN_SUCCESS.

cathay4t replied:
Since this user defined script, they could check NM document to sort it out. Currently, NM does not support this yet.

tyll replied:
What's your opinion of adding support for this in NM?

cathay4t replied:
Reasonable and doable.

tyll commented on examples/user_defined_iface.yml:
```diff
@@ -0,0 +1,43 @@
+    - name: vxcan1
+      type: vxcan
+      description: "user defined vxcan interface"
+      user-defined:
```
When renaming user-defined to generic, maybe this could be called "generic-configuration" or "generic-state".

cathay4t replied:
I do not like `generic`, it is not explicit enough for the use case.

tyll replied:
Another proposal: `plugin`, see above.

tyll commented on examples/user_defined_iface.yml:
```diff
@@ -0,0 +1,43 @@
+user-defined:
+    interface-types:
+        - name: vxcan
+          handler-script: |
```
Just using `handler` might be good enough and allows to simplify the state (no dash) to make it easier to type manually.

cathay4t commented:
The term `plugin` might have different meanings in NetworkManager. I understand the concern, let me find better worlds on this.

liangwen12year commented:
Since github is supported in workflow bot, can you add the jira issue link to the commit message ?

tyll commented on tests/integration/user_define_iface_test.py:
```diff
@@ -0,0 +1,96 @@
+    nm_minor_version() < 46,
+    reason="NetworkManager 1.46- does not support generic device handler",
+)
+def test_user_defined_iface_vxcan():
```
There is a test case missing to create both a vxcan and some other interface at the same time. From a quick glance it looks like the dispatcher script will run for every interface.

cathay4t replied:
Do you mean mixing supported interface type with user defined interface type in single desire state?

NM will make sure dispatch script only been invoked when needed.

cathay4t commented:
New patch uploaded  with new schema.

Instead of forcing user to study backend(NetworkManager) scripts, nmstate is doing simplify it into `activation`, `deactivation` and `allowed-variable-names` to hide the implementation detail.

qinqon commented on examples/dispatch_iface.yml:
```diff
@@ -0,0 +1,21 @@
+---
+dispatch:
+    interface-types:
+        - name: vxcan
+          activation: "ip link add $name type vxcan peer $peer"
+          deactivation: "ip link del $name"
+          allowed-variable-names:
+              - peer
+          kernel-index-getter: "cat /sys/class/net/$name/ifindex"
```
As I understand this correctly a user can apply just the dispatchs types at a yaml and then use them at different yamls.

If this is the case maybe we should allow to delete dispatchs ? adding something like "state: present" or "state: absent"  under interface-types

```yaml
dispatch:
    interface-types:
        - name: vxcan
           state: absent
```

cathay4t replied:
We support so. In commit message we have `To remove a interface type, please set state: absent to interface type in dispatch.interface-types.`

qinqon commented on examples/dispatch_iface.yml:
```diff
@@ -0,0 +1,21 @@
+              - peer
+          kernel-index-getter: "cat /sys/class/net/$name/ifindex"
+interfaces:
+    - name: vxcan1
+      type: dispatch:vxcan
+      description: "vxcan interface created by dispatch script"
+      dispatch:
+          variables:
+              peer: "vxcan1-ep"
+      ipv4:
+          enabled: true
+          address:
+              - ip: 192.0.2.1
+                prefix-length: 24
```
I the unstructured approach at type can be replaced by "name" field under "distpatch" 
```yaml
- name: vxcan1
      type: dispatch
      description: "vxcan interface created by dispatch script"
      dispatch:
          name: vxcan
          variables:
              peer: "vxcan1-ep"
      ipv4:
          enabled: true
          address:
              - ip: 192.0.2.1
                prefix-length: 24
```

cathay4t replied:
A interface need to have a interface type.

qinqon replied:
Can't it be `type: dispatch` ? then we just specify the distpatch type under `distpatch`

cathay4t replied:
User need to look into the `dispatch` section to know the type of this interface seems contradiction to the definition of `interface-type`.

qinqon replied:
Well is like checking if a bridge is stp or not you look at the "bridge" section, the "dispatch:vxcan" is also something that you have to explain to users since it's not something that is clear at the yaml level.

qinqon reviewed (changes requested):
Comments at just the syntax

cathay4t commented:
Changed schema:
 * Interface type changed to `dispatch`
 * For type of dispatch interface, it is required to define as `dispatch.type: vxcan`.

cathay4t commented:
New patch uploaded. Changed `dispatch.interface-type.name` to `dispatch.interfaces.type`.

cathay4t committed:
Add support of interface created by dispatch script

Example YAML:

```yml
dispatch:
  interfaces:
  - type: vxcan
    activation: ip link add $name type vxcan peer $peer
    deactivation: ip link del $name
    kernel-index-getter: cat /sys/class/net/$name/ifindex
    allowed-variable-names:
    - peer
interfaces:
- name: vxcan1
  description: vxcan interface created by dispatch script
  type: dispatch
  state: up
  ipv4:
    enabled: true
    address:
    - ip: 192.0.2.1
      prefix-length: 24
  dispatch:
    type: vxcan
    variables:
      peer: vxcan1-ep
```

To remove a interface type, please set `state: absent` to interface type
in `dispatch.interfaces`.

Example YAML included as `examples/dispatch_iface.yml`.

Integration and unit test cases included.

Signed-off-by: Gris Ge <fge@redhat.com>

liangwen12year commented:
CI `rust_lint` failure indicates that several methods are never used.

cathay4t commented:
Closing because we hear no feedback after we shared demo to the original requestor.
